src/code/z_map_exp.c

```diff
@@ -1,31 +1,255 @@
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A164.s")
+    return -1;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A194.s")
+/**
+ * Returns true if the current scene is a dungeon or boss scene, false otherwise.
+ */
+s32 Map_IsInDungeonOrBossArea(PlayState* play) {
+    if (Map_GetDungeonOrBossAreaIndex(play) == -1) {
+        return false;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A208.s")
+    return true;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A238.s")
+s32 func_8010A0A4(PlayState* play) {
+    if ((Map_GetDungeonOrBossAreaIndex(play) == -1) || (func_80102EF0(play) == 0)) {
+        return false;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A2AC.s")
+    return true;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A2DC.s")
+s32 sDungeonScenes[] = {
+    SCENE_MITURIN, SCENE_HAKUGIN, SCENE_SEA, SCENE_INISIE_N, SCENE_INISIE_R,
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A33C.s")
+/**
+ * If the current scene is a dungeon scene, this function returns an index
+ * indicating which one it is. Otherwise, it returns -1.
+ */
+s32 Map_GetDungeonAreaIndex(PlayState* play) {
+    s32 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A410.s")
+    for (i = 0; i < ARRAY_COUNT(sDungeonScenes); i++) {
+        if (Play_GetOriginalSceneNumber(play->sceneNum) == sDungeonScenes[i]) {
+            return i;
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A430.s")
+    return -1;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A54C.s")
+/**
+ * Returns true if the current scene is a dungeon scene, false otherwise.
+ */
+s32 Map_IsInDungeonArea(PlayState* play) {
+    if (Map_GetDungeonAreaIndex(play) == -1) {
+        return false;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A580.s")
+    return true;
+}
+
+s32 sBossScenes[] = {
+    SCENE_MITURIN_BS, // DUNGEON_INDEX_WOODFALL_TEMPLE
+    SCENE_HAKUGIN_BS, // DUNGEON_INDEX_SNOWHEAD_TEMPLE
+    SCENE_SEA_BS,     // DUNGEON_INDEX_GREAT_BAY_TEMPLE
+    SCENE_INISIE_BS,  // DUNGEON_INDEX_STONE_TOWER_TEMPLE
+};
+
+/**
+ * If the current scene is a boss scene, this function returns an index
+ * indicating which one it is. Otherwise, it returns -1.
+ */
+s32 Map_GetBossAreaIndex(PlayState* play) {
+    s32 i;
+
+    for (i = 0; i < ARRAY_COUNT(sBossScenes); i++) {
+        if (Play_GetOriginalSceneNumber(play->sceneNum) == sBossScenes[i]) {
+            return i;
+        }
+    }
+
+    return -1;
+}
+
+/**
+ * Returns true if the current scene is a boss scene, false otherwise.
+ */
+s32 Map_IsInBossArea(PlayState* play) {
+    if (Map_GetBossAreaIndex(play) == -1) {
+        return false;
+    }
+
+    return true;
+}
+
+s32 D_801BF5A4[] = {
+    SCENE_22DEKUCITY,
+    SCENE_KOEPONARACE,
+    SCENE_F01,
+};
```

Comment:
Any thoughts yet as to what connects these 3 scenes?

Reply:
Nope, we don't know. It's not "scenes that have moving dots on the map", because Stone Tower and Road to Ikana have dots for the boulders. I've also tried just removing all of the things in this array, but there doesn't seem to be any actual effect on the game.